Add error handling when setting the torch mode

shagedorn opened the pull request:
Also fixes a bug, as previously, only `AVCaptureTorchModeOn` was checked for runtime/device support.

This is probably our last PR for now :) Thanks for your responsiveness, really appreciated.

shagedorn commented on Classes/ios/Scanners/MTBBarcodeScanner.m:
```diff
@@ -663,38 +664,61 @@ - (void)updateFocusPreferencesOfDevice:(AVCaptureDevice *)inputDevice reset:(BOO
-- (void)updateTorchModeForCurrentSettings {
+- (BOOL)updateForTorchMode:(MTBTorchMode)preferredTorchMode error:(NSError **)error {
     AVCaptureDevice *backCamera = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
-    if ([backCamera isTorchAvailable] && [backCamera isTorchModeSupported:AVCaptureTorchModeOn]) {
```
I believe this should have always been `[backCamera isTorchModeSupported:[self avTorchModeForMTBTorchMode:self.torchMode]]`

shagedorn commented on Classes/ios/Scanners/MTBBarcodeScanner.m:
```diff
@@ -663,38 +664,61 @@ - (void)updateFocusPreferencesOfDevice:(AVCaptureDevice *)inputDevice reset:(BOO
-        self.torchMode = MTBTorchModeOn;
-    } else {
-        self.torchMode = MTBTorchModeOff;
+    switch (self.torchMode) {
```
This will raise a nice warning, should we ever introduce new torch modes :)

shagedorn commented:
Feel free to let me know should I update any documentation or the sample apps.

mikebuss commented:
Looks good, thanks!

mikebuss commented:
@shagedorn If you get a chance, can you try the Advanced example in the demo project? It looks like the latest changes broke this demo at one point.

mikebuss commented:
It looks like the latest release works in the Advanced example, so the change was in this diff:

https://github.com/mikebuss/MTBBarcodeScanner/compare/master...develop?expand=1

mikebuss commented:
Got it. The scanRect was being set before the session was setup because we're doing it asynchronously now. The updated code is now:

```objective-c
__weak MTBAdvancedExampleViewController *weakSelf = self;
self.scanner.didStartScanningBlock = ^{
    weakSelf.scanner.scanRect = weakSelf.viewOfInterest.frame;
};
```

I'll hold off on releasing this until tomorrow in case you have any feedback @shagedorn.

shagedorn commented:
Will look into it tomorrow morning (German time), but the fix seems reasonable (and necessary).

mikebuss commented:
Thanks, and no rush!

shagedorn commented:
I thought I had fixed this in https://github.com/mikebuss/MTBBarcodeScanner/commit/7e59dbb07d7e23c2fc554ede16d11eb05408f7ec, but looking at `setScanRect:`, it's becomes clear that it cannot work as expected. Thanks for finding the issue.

I have added a safeguard and documentation: https://github.com/mikebuss/MTBBarcodeScanner/pull/109

mikebuss commented:
Thanks! I will release a new version after testing tonight.
